### PR TITLE
Front component

### PR DESCRIPTION
**Pull request by Tabrisrp**
This PR adds the front-side component:
- URLs rewriting with the CDN URL
- Preconnect to the CDN URL in the WP resource hints

### CODE REVIEW

File: src/Front/CDN.php [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+	 *
+	 * @since 1.0
+	 *
+	 * @return string
+	 */
+	private function get_srcset_attributes(): string {
+		/**
+		 * Filter the srcset attributes.
+		 *
+		 * @since 3.8.7
```
Comment (iCaspar) [inline src/Front/CDN.php:235]: Should this be 1.0 (or omit @since, see above)?

File: src/Front/CDN.php [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+	/**
+	 * Search & Replace URLs with the CDN URLs in the provided content
+	 *
+	 * @since 1.0
+	 *
+	 * @param string $html HTML content.
+	 * @return string
+	 */
+	private function rewrite( string $html ): string {
```
Comment (iCaspar) [inline src/Front/CDN.php:98]: The private methods here will be more testable if they can be extracted to a couple other smaller classes:
`Rewriter` class could expose `rewrite()` and `rewrite_srcset()`.
`PathHelper` class could expose the other url/path transformation logic methods.

(The CDN class would be basically a controller without any real logic to test.)

File: src/Front/ServiceProvider.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	public $subscribers = [
+		'cdn_subscriber',
+	];
+
+	/**
+	 * Registers the provided classes
+	 *
+	 * @return void
+	 */
+	public function register() {
```
Comment (iCaspar) [inline src/Front/ServiceProvider.php:33]: If you go with the class extractions above, this would need to be changed to do the injections.

**Review by iCaspar — CHANGES_REQUESTED** [review]
I'm not hard-set on these suggestions, but for consideration/discussion.

File: src/Front/CDN.php (hunk 2 above)
Comment (iCaspar) [reply src/Front/CDN.php:92]: Per discussion, we'll split these out later if we need to.

**Review by iCaspar — APPROVED** [review]

**Comment by piotrbak** [comment]
@Tabrisrp It doesn't look like I have enough permissions to move this PR to the next sprint. Could you move it to 45th?

**Pull request merged**